articles/cubicle-status.md

```diff
@@ -0,0 +1,127 @@
+[here](https://github.com/Sudha247/cubicle/blob/wip-multicore/bwd.ml#L279-L312),
+embedded into this branch of cubicle itself. This has the
+`Functory.Cores.compute` rewritten in Multicore OCaml using Domains and
+Channels.
+
+### Observations
+
+We modified cubicle to use the multicore interface of functory, the code is
+available in this [branch](https://github.com/Sudha247/cubicle/tree/wip-multicore).
+
+It was causing segmentation faults while executing on multiple domains,
+possibly due to multiple domains altering the same objects simultaneously.
+Segfaults have been resolved in 4.10.0+multicore compiler variant(parallel
+minor GC) and we are able to obtain the backtrace of the failures.
+
+For instance this is the backtrace for `german_pfs` while executing on 4 cores:
+
+```
+$ ./cubicle.opt examples/german_pfs.cub -j 4
+cores: 4
+
+node 1: unsafe[1]
+                                                                                                    39 (24+15) remaining
+
+
+Not_found
+Fatal error: Not_found
+Raised at file "common/vec.ml", line 74, characters 23-38
+Called from file "common/iheap.ml", line 73, characters 44-63
+Called from file "common/iheap.ml", line 102, characters 9-22
+Called from file "list.ml", line 110, characters 12-15
+Called from file "list.ml", line 110, characters 12-15
+Called from file "smt/solver.ml", line 945, characters 2-123
+Called from file "smt/alt_ergo.ml", line 538, characters 6-27
+Called from file "prover.ml", line 215, characters 2-61
+Called from file "prover.ml", line 226, characters 2-24
+Called from file "fixpoint.ml", line 336, characters 4-18
+Called from file "fixpoint.ml", line 390, characters 6-28
+Called from file "bwd.ml", line 199, characters 15-47
+Called from file "bwd.ml" (inlined), line 332, characters 18-37
+Called from file "bwd.ml", line 300, characters 29-39
+Called from file "bwd.ml", line 307, characters 6-26
+```
+
+### Problem
+
+The backward reachability algorithm uses the same instance of the SMT solver
+for the entire search. This doesn't go well with Multicore OCaml because all
+domains operate on the same major heap, with possible operations on objects by
+multiple domains (analogous to threads) simultaneously.
+
+One solution to this is to have a separate instance of the SMT solver
+for each domain. This way one domain does not meddle with data of another
+domain.
+
+>Cubicle’s integration with the SMT solver at the API level is crucial for
+efficient treatment of the subsumption check. For any such check, a single
+context for the SMT solver is used; it just gets incremented and repeatedly
+verified. To support the efficient (symmetry-reduced) and exhaustive
+application of the inexpensive redundancy and subset checks, cubes are
+maintained in normal form where variables are renamed and implied literals
+removed at construction time.
+
+-- [1]
+
+This statement although poses the question whether the correctness of the BRAB
```

Comment:
Is this what cubicle already does when it creates multiple processes using functory? Or is there something else happening that allows sharing the BFS between multiple processes?

Reply:
IIUC `Unix.fork` copies the entire heap onto the new process at the time of creation, but later on the when the process continues the data alterations it makes is local to itself. If that is the case, then this might not be very different from having a per-domain SMT instance and that approach should work on multiple domains, though I'm not sure as yet if the SMT solver itself has any global state that might be affected if we make multiple instances of it. I'll investigate this further and post any updates here.

Reply:
> IIUC Unix.fork copies the entire heap onto the new process at the time of creation, but later on the when the process continues the data alterations it makes is local to itself.

Other than the subtlety that pages are copied-on-write, that is correct. I expect that having the data structures per domain should work.